straight-thaw-versions ergonomics improvements: auto-fetch, no decapitation

antifuchs opened the pull request:
This PR is kinda a catch-all for the things that I found that were ... not ideal with thawing:

* #294 - freezing after thawing is not a no-op
* #66 - thawing leaves detached HEADs around
* #58 - thawing requires fetching beforehand

I think the attached change solves all of these in a fairly neat way,
and I hope you think so too (and that I didn't miss some subtle
interaction!).

`straight-vc-git-check-out-commit` now normalizes the local repo when checking out a commit.

`straight-thaw-versions` now:

* Tries to ensure that the commit that we're thawing to is present by checking it's there, and if not, fetching.
* Uses a new and revamped check-out-commit function that `resets --hard` instead of running `checkout`.

With this change, I managed to thaw and re-freeze the versions (with an updated repro case from #294), and they were stable!

I've also updated the README to not mention having to fetch before a freeze, as freeze does that automatically now.

Hope this is reasonable!

antifuchs commented:
Phew, took me 2 tries to get lints to work out, but I think this should do it!

raxod502 commented on straight.el:
```diff
@@ -1212,6 +1212,14 @@ repository directory and delegates to the relevant
+      (local-repo type)
+    (let ((straight--default-directory (straight--repos-dir local-repo)))
+      (straight-vc 'commit-present-p type local-repo commit))))
+
```
This needs a bullet point in the README section that lists VC backend operations, as well as one in the section that describes the Git backend's implementations of them.

raxod502 commented on straight.el:
```diff
@@ -2066,9 +2074,17 @@ is a 40-character string identifying a Git commit."
+           (straight--get-call "git" "reset" "--hard" commit)
            (cl-return)))))
 
+(cl-defun straight-vc-git-commit-present-p (local-repo commit)
```
This should have a docstring.

raxod502 commented on straight.el:
```diff
@@ -4987,6 +5003,9 @@ according to the value of `straight-profiles'."
              (when-let ((commit (cdr (assoc local-repo versions-alist))))
+               (straight-vc-normalize recipe)
+               (unless (straight-vc-commit-present-p recipe commit)
+                 (straight-vc-fetch-from-upstream recipe))
```
I think we should also have a `straight-vc-fetch-from-remote`, no?

antifuchs replied:
Oh myyyy, yes it should be that, I missed that!

raxod502 commented on straight.el:
```diff
@@ -2066,9 +2074,17 @@ is a 40-character string identifying a Git commit."
+(cl-defun straight-vc-git-commit-present-p (local-repo commit)
+  (cl-block nil
+    (while t
+      (and (straight-vc-git--ensure-nothing-in-progress local-repo)
```
Do we need this?

antifuchs replied:
I wasn't sure - it seemed like a good idea to prevent concurrent operations that the user might have kicked off inadvertently, but given that this is non-destructive, it should be extremely ok leave the looping/waiting to functions that are destructive. I'll remove it.

raxod502 replied:
I think my position is that if you run Git commands concurrently with `straight.el` VC operation, then you are perfectly well entitled to shoot yourself in the foot. It would be hopeless to try to fix all the cases that would result from such unpredictable state.

antifuchs commented:
Thanks so much for the review! I have gone through your suggestions & think I addressed them! 

I'd be excited for this to be part of the NEWS - we should probably do this as part of another commit, since I don't yet know when it'll go in (:

raxod502 commented on straight.el:
```diff
@@ -2066,9 +2074,14 @@ is a 40-character string identifying a Git commit."
       (and (straight-vc-git--ensure-nothing-in-progress local-repo)
            (straight-vc-git--ensure-worktree local-repo)
-           (straight--get-call "git" "checkout" commit)
+           (straight--get-call "git" "reset" "--hard" commit)
```
Now that I'm looking at this again, it seems to me that the normalization logic should be included here. When would we ever want to check out a commit by resetting a branch other than the primary branch?

The fetch, though, should probably happen where you have it.

antifuchs commented:
Alright - I moved this to the `develop` branch, made that News entry (dating it May 1; feel free to adjust if you need), and added the normalization logic to the check-out-commit function. That needed an incompatible change to the vc interface: check-out-commit only took the local repo, not the recipe (which is needed to normalize). 

I adjusted the interfaces, but didn't do anything else about it - please let me know if there's anything else that I should do about it (:

raxod502 commented on README.md:
```diff
@@ -2682,6 +2685,14 @@ with Emacs, rather than cloning the upstream repository:
 [Read more.][#user/recipes]
 
 ## News
+### May 1, 2019
```
Run `make toc` to update the table of contents.

raxod502 commented on straight.el:
```diff
@@ -1203,8 +1203,8 @@ repository directory and delegates to the relevant
-  "Using VC backend TYPE, in LOCAL-REPO, check out COMMIT.
+(defun straight-vc-check-out-commit (recipe commit)
+  "Normalize the repo for RECIPE and check out COMMIT.
 TYPE is a symbol like symbol `git', etc. LOCAL-REPO is a string
```
This docstring needs updating.

antifuchs replied:
While I was there, I adjusted the function's behavior to not do anything if there's no local-repo, symmetrical to the other vc functions that take a recipe.

antifuchs commented:
Updated again (:

antifuchs commented:
\o/ - thanks for the thoughtful reviews! I'm super excited to use this going forward (:
